qualtran/bloqs/rotations/phase_gradient.py

```diff
@@ -520,21 +522,32 @@ def apply(
     ) -> Tuple[
         Union[int, np.integer, NDArray[np.integer]], Union[int, np.integer, NDArray[np.integer]]
     ]:
-        out = self.on_classical_vals(x=x, phase_grad=phase_grad)
-        return out['x'], out['phase_grad']
+        phase_grad_out = (phase_grad + self.scaled_val_int(x)) % 2**self.phase_bitsize
+        return x, phase_grad_out
+
+    def on_classical_vals(self, x: Fxp, phase_grad: Fxp) -> Dict[str, Fxp]:
+        if is_symbolic(self.phase_bitsize):
+            raise ValueError(f"Cannot classically simulate with symbolic {self.phase_bitsize=}")
+        if is_symbolic(self.gamma):
+            raise ValueError(f"Cannot classically simulate with symbolic {self.gamma=}")
 
-    def on_classical_vals(self, x: int, phase_grad: int) -> Dict[str, 'ClassicalValT']:
-        phase_grad_out = (phase_grad + self.scaled_val(x)) % 2**self.phase_bitsize
+        scaled_x = _mul_via_repeated_add(
+            x.like(phase_grad), gamma_fxp=self.abs_gamma_fxp, out=self.phase_bitsize
```

Comment:
I think it's wrong to do the `x.like(phase_grad)` truncation here. `x` can be bigger than phase_grad and in this case, we should not discard the extra bits in `x`. The `_mul_via_repeated_add` will add shifted substrings of `x` into `phase_grad` but if we discard the extra bits here then the shifted substrings would just be `0` and lead to a larger approximation error. 

Ideally, a test should be failing at this point

Reply:
Ah you're right. All the current tests have `x.n_int = 0`. I'll add a test to catch this and fix it

Reply:
The issue is not necessarily because of `n_int`; but also because the fractional bitsize can be larger than `phase_grad` and if gamma > 1; then you'd be doing a left shift and adding into phase_grad; so you'll add bits 5, 6 7 into a phase grad register of size 3 if left shift due to gamma is 4